io/src/lib.rs

```diff
@@ -0,0 +1,225 @@
+pub enum GameStage {
+    #[default]
+    Preparation,
+    InProgress(StageDescription),
+    Reveal(StageDescription),
+}
+
+impl GameStage {
+    pub fn game_is_in_progress(&self) -> bool {
+        match self {
+            GameStage::Preparation => false,
+            GameStage::InProgress(_) | GameStage::Reveal(_) => true,
+        }
+    }
+
+    pub fn move_can_be_made(&self) -> bool {
+        match self {
+            GameStage::Preparation | GameStage::InProgress(_) => true,
+            GameStage::Reveal(_) => false,
+        }
+    }
+
+    pub fn is_player_in_game(&self, player: &ActorId) -> bool {
+        match self {
+            GameStage::Preparation => false,
+            GameStage::InProgress(description) | GameStage::Reveal(description) => {
+                description.anticipated_players.contains(player)
+                    || description.finished_players.contains(player)
+            }
+        }
+    }
+
+    pub fn current_players(&self) -> Option<BTreeSet<ActorId>> {
+        let description = match self {
+            GameStage::Preparation => return None,
+            GameStage::InProgress(progress_description) => progress_description,
+            GameStage::Reveal(reveal_description) => reveal_description,
+        };
+
+        let players = description
+            .anticipated_players
+            .union(&description.finished_players)
+            .copied()
+            .collect();
+        Some(players)
+    }
+}
+
+#[derive(Debug, Encode, Decode, TypeInfo)]
+pub struct Duration {
+    pub days: u64,
+    pub hours: u64,
+    pub minutes: u64,
+}
+
+#[derive(Debug, Encode, Decode, TypeInfo)]
+pub enum RevealResult {
+    Continue,
+    NextRoundStarted { players: BTreeSet<ActorId> },
+    GameOver { winner: ActorId },
+}
+
+#[derive(Debug, Decode, Encode, TypeInfo)]
+pub enum Action {
+    /// Registers a player for the game.
+    /// Player must send value to be registered
+    ///
+    /// # Requirements:
+    /// * Game is not in progress yet. E.g. the `GameStage` must be `GameStage::Preparation`
+    /// * `msg::value()` is greater or equal to `bet_size` in the config(refund will return to user).
+    /// * Player not registred yet.
+    /// * Lobby is not full.
+    ///
+    /// On success replies `Event::PlayerRegistred`.
+    Register,
+
+    /// Submits player's move to the program in encrypted form.
+    /// Player can't change his move after it.
+    ///
+    /// # Arguments:
+    /// * `String`: is the hex string from 256-bit blake2b hash of move("0" or "1" or "2" or "3" or "4") + "password".
+    ///
+    /// # Requirements:
+    /// * The `GameStage` must be `GameStage::InProgress(StageDesciption)` where `StageDescription::anticipated_players` must contains `msg::source()`
+    ///
+    /// On success replies `Event::SuccessfulReveal(RevealResult)` where `RevealResult` will correspond to the situation after this reveal.
+    MakeMove(String),
```

Comment:
I don't see the reason to use thee UTF-8 `String` here.
```suggestion
    MakeMove(Vec<u8>),
```